README.md
# Spartaaa

[![Build Status](https://travis-ci.com/codekie/spartaaa.svg?branch=master)](https://travis-ci.com/codekie/spartaaa)
[![Coverage Status](https://coveralls.io/repos/github/codekie/spartaaa/badge.svg?branch=master)](https://coveralls.io/github/codekie/spartaaa?branch=master) [![Greenkeeper badge](https://badges.greenkeeper.io/codekie/spartaaa.svg)](https://greenkeeper.io/)

`Spartaaa` is a Web-frontend for taskwarrior.

**CAUTION: USE AT OWN RISK!** If you use this app, make sure that you made a backup of the Taskwarrior-data, first.


# Requirements

- `Taskwarrior` pre-installed
- NodeJS `8`
- macOS

If the app is not run on a `macOS`, the path to the Taskwarrior-data, has to be set as env-variable
(`SPARTAAA_PATH_TASKWARRIOR_DATA`).


# How to run the app

The dependencies, for the server, have to be installed first. To do this, run

```
yarn install
```

To start the server, execute

```
yarn start
```

The web-app will be accessible at http://localhost:3010


# How does it work

While the web-app itself, is served from an express-webserver, the communication between frontend and backend happens
via Websockets.

This app does not access the Taskwarrior-data directly. The access happens through the Taskwarrior-CLI.

There is a watch on the directory, which contains the Taskwarrior-data (preconfigured for `macOS`, which is `~/.task`)
and as soon as any file within that directory is modified, the tasks will be exported and pushed to the client.


# Development

## How to start the tests

```
yarn test
```


## Start the server in dev mode

```
yarn run start-dev
```

This will start the express- and websocket-server, additionally to a Webpack-dev-server. There are file-watches on
the server-components, which will restart the server automatically, as well as file-watches on the frontend-code, which
will recompile the frontend and hot-reload it in the browser.

The Webpack-dev-server, which will serve the hot-reloadable frontend, will be accessible at http://localhost:8080


## Motivation

The initial motivation, was to create a React-boilerplate for future projects. Taskwarrior was just a proof of concept
to build upon. However, it turned to out to become useful and so the development went on.

## Used technologies / frameworks

- Yarn
- React
- Redux
- immutable.js
- RxJS
- SASS
- Bulma
- Fontawesome
- Webpack
- Jest
- Babel
- Express
- lodash
- moment.js
- winston
- ws


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.0"></a>
# 0.2.0 (2018-10-31)


### Bug Fixes

* restore hot-update functionality ([b1ef6fc](https://github.com/codekie/spartaaa/commit/b1ef6fc))


### Features

* add 'complete' TaskList-view ([15ef194](https://github.com/codekie/spartaaa/commit/15ef194))
* add "activate"- and "deactivate"-button ([3d6095b](https://github.com/codekie/spartaaa/commit/3d6095b))
* add "postponed"-view ([08e84ae](https://github.com/codekie/spartaaa/commit/08e84ae))
* add buttons for finishing and "unfinishing" a task ([769f50b](https://github.com/codekie/spartaaa/commit/769f50b))
* add custom sorting for TaskList-views ([c0cf40a](https://github.com/codekie/spartaaa/commit/c0cf40a))
* add favicon ([c727ab1](https://github.com/codekie/spartaaa/commit/c727ab1))
* add project-tag to task ([a8dc739](https://github.com/codekie/spartaaa/commit/a8dc739))
* automatically reconnect to WebSocket-server, upon disconnect ([d2d21ff](https://github.com/codekie/spartaaa/commit/d2d21ff))
* enable hot module replacement ([d631520](https://github.com/codekie/spartaaa/commit/d631520))
* implement task-filter with `WebSocket`-session ([7998b0f](https://github.com/codekie/spartaaa/commit/7998b0f))
* introduce `TaskListView`s ([92862c8](https://github.com/codekie/spartaaa/commit/92862c8))
* push changes to client on file-changes ([0ef402d](https://github.com/codekie/spartaaa/commit/0ef402d))
* show `due` since/until ([ef0b7fe](https://github.com/codekie/spartaaa/commit/ef0b7fe))
* show connection state ([8e30258](https://github.com/codekie/spartaaa/commit/8e30258))
* show priority as "bar" instead of colorized text ([83fb9a3](https://github.com/codekie/spartaaa/commit/83fb9a3))
* show uuid, instead of id, on completed tasks ([ae228f9](https://github.com/codekie/spartaaa/commit/ae228f9))
* split app and vendor code ([73d2aaf](https://github.com/codekie/spartaaa/commit/73d2aaf))
* use icons for group-tags ([1fc9c7c](https://github.com/codekie/spartaaa/commit/1fc9c7c))
* use icons for tasks ([cc014b5](https://github.com/codekie/spartaaa/commit/cc014b5))
* use redux for state-management ([872f76a](https://github.com/codekie/spartaaa/commit/872f76a))
* use websockets for communication ([43e3f77](https://github.com/codekie/spartaaa/commit/43e3f77))


config/app.js
const path = require('path'),
    os = require('os');

const DEFAULT__PORT__DEV_SERVER = 8080,
    DEFAULT__PORT__SERVER = 3010,
    DEFAULT__PORT__SERVER__WEBSOCKET = 40520,
    DEFAULT__PORT__TEST__SERVER = 3020,
    DEFAULT__PROTOCOL__SERVER = 'http',
    DEFAULT__PROTOCOL__SERVER__WEBSOCKET = 'ws',
    DEFAULT__HOSTNAME__SERVER = 'localhost',
    DEFAULT__PATH__TASKWARRIOR_DATA = `${ os.homedir() }/.task`,
    DEFAULT__FILE_WATCH__DEBOUNCE_DURATION = 20,
    ENV_PREFIX = 'SPARTAAA_',
    ENV__PORT__SERVER = `${ ENV_PREFIX }PORT`,
    ENV__PORT__SERVER__WEBSOCKET = `${ ENV_PREFIX }PORT_WEBSOCKET`,
    ENV__PORT__DEV__SERVER = `${ ENV_PREFIX }PORT_DEV`,
    ENV__PATH__TASKWARRIOR_DATA = `${ ENV_PREFIX }PATH_TASKWARRIOR_DATA`,
    BASE_PATH__SERVICES = '/api',
    PATH__PROJECT_BASE = path.join(__dirname, '..'),
    PATH__BUILD_OUTPUT = path.join(PATH__PROJECT_BASE, 'dist'),
    PATH__SERVER__STATIC_FILES = path.join(PATH__PROJECT_BASE, 'src/server/static'),
    FILEPATH__DEV__ENTRY__CLIENT = path.join(PATH__PROJECT_BASE, 'src/client/index.jsx'),
    FILEPATH__BUILD__CLIENT__TEMPLATE = path.join(PATH__PROJECT_BASE, 'src/client/index.html'),
    FILEPATH__OUTPUT__CLIENT__TEMPLATE = path.join(PATH__BUILD_OUTPUT, 'index.html'),
    FILENAME__ENTRY__SERVER = 'server.js';

const config = _createConfig();

// TODO log config to console, in DEV-mode

module.exports = config;

function _createConfig() {
    return {
        entry: _createEntryConfig(),
        server: _createServerConfig(),
        dev: _createDevConfig(),
        test: _createTestConfig(),
        build: _createBuildConfig()
    };
}

function _createEntryConfig() {
    return {
        server: path.join(PATH__BUILD_OUTPUT, FILENAME__ENTRY__SERVER)
    };
}

function _createServerConfig() {
    const port = process.env[ENV__PORT__SERVER] || DEFAULT__PORT__SERVER,
        portWebSocket = process.env[ENV__PORT__SERVER__WEBSOCKET] || DEFAULT__PORT__SERVER__WEBSOCKET;
    return {
        port,
        host: `${ DEFAULT__PROTOCOL__SERVER }://${ DEFAULT__HOSTNAME__SERVER }:${ port }`,
        basePathServices: BASE_PATH__SERVICES,
        pathStaticFiles: PATH__SERVER__STATIC_FILES,
        pathTaskwarriorData: process.env[ENV__PATH__TASKWARRIOR_DATA] || DEFAULT__PATH__TASKWARRIOR_DATA,
        fileWatch: {
            debounceDuration: DEFAULT__FILE_WATCH__DEBOUNCE_DURATION
        },
        websocket: {
            port: portWebSocket,
            host: `${ DEFAULT__PROTOCOL__SERVER__WEBSOCKET }://${ DEFAULT__HOSTNAME__SERVER }:${ portWebSocket }`
        }
    };
}

function _createDevConfig() {
    return {
        client: {
            filepathEntry: FILEPATH__DEV__ENTRY__CLIENT,
            devServer: {
                port: process.env[ENV__PORT__DEV__SERVER] || DEFAULT__PORT__DEV_SERVER
            }
        }
    };
}

function _createTestConfig() {
    return {
        server: {
            port: DEFAULT__PORT__TEST__SERVER
        }
    };
}

function _createBuildConfig() {
    return {
        outputPath: PATH__BUILD_OUTPUT,
        client: {
            input: {
                filepathTemplate: FILEPATH__BUILD__CLIENT__TEMPLATE
            },
            output: {
                filepathTemplate: FILEPATH__OUTPUT__CLIENT__TEMPLATE,
                path: PATH__BUILD_OUTPUT
            }
        },
        server: {
            outputFilename: FILENAME__ENTRY__SERVER
        }
    };
}
